### ISSUE TITLE
relax 80 col limit and switch to semantic line breaks

### ISSUE THREAD

**Issue by timotheecour** [timotheecour/Nim#649]
## proposal
* use semantic line breaks, ie split at logical boundaries (commas, end of sentence etc), not at 80 column limit which causes reflow every time you add a word to a sentence in a comment etc
* likewise in code

## example
avoid need-lessly splitting in cases like this:
https://github.com/nim-lang/Nim/pull/17338

```nim
    let output8 = input8.toHtml(error=error8)
    check(error8[] == "input(1, 1) Error: new section expected (overline " &
            "\'-----------\' is too short)")
```

## links
* [sembr/specification: Semantic Line Breaks Specification](https://github.com/sembr/specification)
* [Semantic Linebreaks - Kristian Glass - Do I Smell Burning?](https://blog.doismellburning.co.uk/semantic-linebreaks/)
* [Semantic Linefeeds](https://rhodesmill.org/brandon/2012/one-sentence-per-line/)
* [Semantic Line Breaks](https://sembr.org/)
* [Semantic line breaks | Ramshankar](https://ramshankar.org/blog/posts/2019/semantic-line-breaks/)
* [Ask Hackaday: Are 80 Characters Per Line Still Reasonable In 2020? | Hackaday](https://hackaday.com/2020/06/18/ask-hackaday-are-80-characters-per-line-still-reasonable-in-2020/)
* [One sentence per line, please | Hacker News](https://news.ycombinator.com/item?id=4642395)
* [nimpretty should keep proc declaration as 1 line instead of breaking into multiple lines · Issue #189 · nim-lang/RFCs](https://github.com/nim-lang/RFCs/issues/189)

**Comment by a-mr** [comment]
This can be improved by just introducing a new function, say `oneline` (like it was done for multi-line case in `dedent`):

```
    check(error8[] == oneline"""input(1, 1) Error: new section expected
                                (overline \'-----------\' is too short)""")
```
assuming the rule that any "line break in the string + many spaces" is converted into exactly one space.

It's because I support 80-character limit... May be in the ideal world by year 2100 we will be ready for this proposal...

**Comment by timotheecour** [comment]
`online` creates further complications, in fact even the example you show demonstrates this: it makes `unittest.check` slightly worse by making it example not just the lhs but also the rhs since it becomes a non-litteral. All this being avoidable by relaxing the 80 col rule.

But this proposal has 2 (mostly orthogonal) parts, do you at least agree with 1st part, which states that line breaks should be semantics rather than upon reaching the column limit:
```nim
## The compiler depends on the System module to work properly and the System
## module depends on the compiler. Most of the routines listed here use
## special compiler magic.
##
## Each module implicitly imports the System module; it must not be listed
## explicitly. Because of this there cannot be a user-defined module named
## `system`.
##
```

=>

```nim
## The compiler depends on the System module to work properly and
## the System module depends on the compiler.
## Most of the routines listed here use special compiler magic.
##
## Each module implicitly imports the System module;
## it must not be listed explicitly.
## Because of this there cannot be a user-defined module named `system`.
```

The rendered html is the same, but the source is easier to maintain, eg in case you add a word in 1 sentence/proposition, you typically don't need to do any reflowing. Thus it is also more diff friendly.

Whether to relax the 80 col limit (eg 80 soft, 120 hard) is a separate issue from this.

**Comment by a-mr** [comment]
Yes, that part I agree with. Actually I had very similar thoughts about doc comments, but I did not even think that someone had already come up with a smart theory about that.

I have a feeling that this semantic line breaks should be phrased into some simple algorithm without smart rantings :-), like:

- if the end of phrase with `.` or `;` is reached then break the line
- if 80 characters limit is reached then go back on this line to one of `and`, `or`, `while`, `then`, `which`, `therefore` and break the line there

so that people can format text **fast** and **mindlessly** (it's a good thing here).

Regarding the value of limit... I once read the book that was really impossible to read with a decent speed because of **extremely** long lines. Can you guess how many characters were in the line of that book? Yes, 80. Generally comfortable line width for reading is not more than 60-70 characters. It's the basics of printing and they follow it in the books fortunately. So 80 characters may be a good value for the **hard** limit.
